From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Sun, 12 Mar 2023 08:21:56 +0800
Subject: [PATCH] Add ChunkHot from KioCGServer


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index eccc357cabd8317afc5666c3347936d50eb34015..19e16ce3886b34ea506d6adeae92db8589bf7073 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -291,4 +291,6 @@ public interface Chunk extends PersistentDataHolder {
      * @return if the biome is contained within
      */
     boolean contains(@NotNull Biome biome);
+
+    long getChunkHot(); // KioCG
 }
