From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Wed, 25 Jan 2023 21:27:50 +0800
Subject: [PATCH] Rebrand again


diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index dd5f91a3bbb3a6bef1d8e05bdbb5ceb6a8259f55..c83e6dbf68c1f0997e797fc43ad3d5c909f49460 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -20,7 +20,7 @@ import java.util.stream.StreamSupport;
 public class PaperVersionFetcher implements VersionFetcher {
     private static final java.util.regex.Pattern VER_PATTERN = java.util.regex.Pattern.compile("^([0-9\\.]*)\\-.*R"); // R is an anchor, will always give '-R' at end
     // Purpur start
-    private static final String DOWNLOAD_PAGE = "https://github.com/NaturalCodeClub/HearseRewrite/releases/tag/Auto-Build";
+    private static final String DOWNLOAD_PAGE = "https://github.com/Era4FunMC/Hearse/releases/tag/Auto-Build";
     private static int distance = -2; public int distance() { return distance; }
     // Purpur end
     private static @Nullable String mcVer;
@@ -34,7 +34,7 @@ public class PaperVersionFetcher implements VersionFetcher {
     @Override
     public Component getVersionMessage(@Nonnull String serverVersion) {
         String[] parts = serverVersion.substring("git-Hearse-".length()).split("[-\\s]"); // Purpur //Hearse
-        final Component updateMessage = getUpdateStatusMessage("NaturalCodeClub/HearseRewrite", "ver/" + getMinecraftVersion(), parts[0]); // Purpur //Hearse
+        final Component updateMessage = getUpdateStatusMessage("Era4FunMC/Hearse", "ver/" + getMinecraftVersion(), parts[0]); // Purpur //Hearse
         final Component history = getHistory();
 
         return history != null ? Component.join(net.kyori.adventure.text.JoinConfiguration.separator(Component.newline()), history, updateMessage) : updateMessage; // Purpur
