From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Fri, 3 Feb 2023 15:12:21 +0800
Subject: [PATCH] Tick players in main thread


diff --git a/src/main/java/co/earthme/hearse/server/ServerEntityTickHook.java b/src/main/java/co/earthme/hearse/server/ServerEntityTickHook.java
index 891a9167e4072d6ce5b526b1676d8ecb836c8733..b34e7432065f8101c1c4ad85e85cf932639f2b75 100644
--- a/src/main/java/co/earthme/hearse/server/ServerEntityTickHook.java
+++ b/src/main/java/co/earthme/hearse/server/ServerEntityTickHook.java
@@ -8,6 +8,7 @@ import co.earthme.hearse.concurrent.threadfactory.DefaultWorkerFactory;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import java.util.concurrent.LinkedBlockingQueue;
@@ -75,7 +76,7 @@ public class ServerEntityTickHook {
                 taskCounter.getAndDecrement();
             }
         };
-        if (!asyncEntityEnabled){
+        if (!asyncEntityEnabled || entity instanceof Player){
             task.run();
             return;
         }
